// @playwright/mcp/lib/tools/snapshot.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const zod_1 = require("zod");
const zod_to_json_schema_1 = __importDefault(require("zod-to-json-schema"));
const context_1 = require("../context");
const javascript = __importStar(require("../javascript"));
const snapshot = {
    capability: 'core',
    schema: {
        name: 'browser_snapshot',
        description: 'Capture accessibility snapshot of the current page, this is better than screenshot',
        inputSchema: (0, zod_to_json_schema_1.default)(zod_1.z.object({})),
    },
    handle: async (context) => {
        const tab = await context.ensureTab();
        return await tab.run(async () => {
            const code = [`// <internal code to capture accessibility snapshot>`];
            return { code };
        }, { captureSnapshot: true });
    },
};
const elementSchema = zod_1.z.object({
    element: zod_1.z.string().describe('Human-readable element description used to obtain permission to interact with the element'),
    ref: zod_1.z.string().describe('Exact target element reference from the page snapshot'),
});
const click = {
    capability: 'core',
    schema: {
        name: 'browser_click',
        description: 'Perform click on a web page',
        inputSchema: (0, zod_to_json_schema_1.default)(elementSchema),
    },
    handle: async (context, params) => {
        const validatedParams = elementSchema.parse(params);
        return await context.currentTab().runAndWaitWithSnapshot(async (snapshot) => {
            const locator = snapshot.refLocator(validatedParams.ref);
            const code = [
                `// Click ${validatedParams.element}`,
                `await page.${await (0, context_1.generateLocator)(locator)}.click();`
            ];
            await locator.click();
            return { code };
        });
    },
};
const dragSchema = zod_1.z.object({
    startElement: zod_1.z.string().describe('Human-readable source element description used to obtain the permission to interact with the element'),
    startRef: zod_1.z.string().describe('Exact source element reference from the page snapshot'),
    endElement: zod_1.z.string().describe('Human-readable target element description used to obtain the permission to interact with the element'),
    endRef: zod_1.z.string().describe('Exact target element reference from the page snapshot'),
});
const drag = {
    capability: 'core',
    schema: {
        name: 'browser_drag',
        description: 'Perform drag and drop between two elements',
        inputSchema: (0, zod_to_json_schema_1.default)(dragSchema),
    },
    handle: async (context, params) => {
        const validatedParams = dragSchema.parse(params);
        return await context.currentTab().runAndWaitWithSnapshot(async (snapshot) => {
            const startLocator = snapshot.refLocator(validatedParams.startRef);
            const endLocator = snapshot.refLocator(validatedParams.endRef);
            const code = [
                `// Drag ${validatedParams.startElement} to ${validatedParams.endElement}`,
                `await page.${await (0, context_1.generateLocator)(startLocator)}.dragTo(page.${await (0, context_1.generateLocator)(endLocator)});`
            ];
            await startLocator.dragTo(endLocator);
            return { code };
        });
    },
};
const hover = {
    capability: 'core',
    schema: {
        name: 'browser_hover',
        description: 'Hover over element on page',
        inputSchema: (0, zod_to_json_schema_1.default)(elementSchema),
    },
    handle: async (context, params) => {
        const validatedParams = elementSchema.parse(params);
        return await context.currentTab().runAndWaitWithSnapshot(async (snapshot) => {
            const locator = snapshot.refLocator(validatedParams.ref);
            const code = [
                `// Hover over ${validatedParams.element}`,
                `await page.${await (0, context_1.generateLocator)(locator)}.hover();`
            ];
            await locator.hover();
            return { code };
        });
    },
};
const typeSchema = elementSchema.extend({
    text: zod_1.z.string().describe('Text to type into the element'),
    submit: zod_1.z.boolean().optional().describe('Whether to submit entered text (press Enter after)'),
    slowly: zod_1.z.boolean().optional().describe('Whether to type one character at a time. Useful for triggering key handlers in the page. By default entire text is filled in at once.'),
});
const type = {
    capability: 'core',
    schema: {
        name: 'browser_type',
        description: 'Type text into editable element',
        inputSchema: (0, zod_to_json_schema_1.default)(typeSchema),
    },
    handle: async (context, params) => {
        const validatedParams = typeSchema.parse(params);
        return await context.currentTab().runAndWaitWithSnapshot(async (snapshot) => {
            const locator = snapshot.refLocator(validatedParams.ref);
            const code = [];
            if (validatedParams.slowly) {
                code.push(`// Press "${validatedParams.text}" sequentially into "${validatedParams.element}"`);
                code.push(`await page.${await (0, context_1.generateLocator)(locator)}.pressSequentially(${javascript.quote(validatedParams.text)});`);
                await locator.pressSequentially(validatedParams.text);
            }
            else {
                code.push(`// Fill "${validatedParams.text}" into "${validatedParams.element}"`);
                code.push(`await page.${await (0, context_1.generateLocator)(locator)}.fill(${javascript.quote(validatedParams.text)});`);
                await locator.fill(validatedParams.text);
            }
            if (validatedParams.submit) {
                code.push(`// Submit text`);
                code.push(`await page.${await (0, context_1.generateLocator)(locator)}.press('Enter');`);
                await locator.press('Enter');
            }
            return { code };
        });
    },
};
const selectOptionSchema = elementSchema.extend({
    values: zod_1.z.array(zod_1.z.string()).describe('Array of values to select in the dropdown. This can be a single value or multiple values.'),
});
const selectOption = {
    capability: 'core',
    schema: {
        name: 'browser_select_option',
        description: 'Select an option in a dropdown',
        inputSchema: (0, zod_to_json_schema_1.default)(selectOptionSchema),
    },
    handle: async (context, params) => {
        const validatedParams = selectOptionSchema.parse(params);
        return await context.currentTab().runAndWaitWithSnapshot(async (snapshot) => {
            const locator = snapshot.refLocator(validatedParams.ref);
            const code = [
                `// Select options [${validatedParams.values.join(', ')}] in ${validatedParams.element}`,
                `await page.${await (0, context_1.generateLocator)(locator)}.selectOption(${javascript.formatObject(validatedParams.values)});`
            ];
            await locator.selectOption(validatedParams.values);
            return { code };
        });
    },
};
const screenshotSchema = zod_1.z.object({
    raw: zod_1.z.boolean().optional().describe('Whether to return without compression (in PNG format). Default is false, which returns a JPEG image.'),
});
const screenshot = {
    capability: 'core',
    schema: {
        name: 'browser_take_screenshot',
        description: `Take a screenshot of the current page. You can't perform actions based on the screenshot, use browser_snapshot for actions.`,
        inputSchema: (0, zod_to_json_schema_1.default)(screenshotSchema),
    },
    handle: async (context, params) => {
        const validatedParams = screenshotSchema.parse(params);
        const tab = context.currentTab();
        const options = validatedParams.raw ? { type: 'png', scale: 'css' } : { type: 'jpeg', quality: 50, scale: 'css' };
        const screenshot = await tab.page.screenshot(options);
        return {
            content: [{ type: 'image', data: screenshot.toString('base64'), mimeType: validatedParams.raw ? 'image/png' : 'image/jpeg' }],
        };
    },
};
exports.default = [
    snapshot,
    click,
    drag,
    hover,
    type,
    selectOption,
    screenshot,
];
