// @playwright/mcp/lib/tools/tabs.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const zod_1 = require("zod");
const zod_to_json_schema_1 = require("zod-to-json-schema");
const listTabs = {
    capability: 'tabs',
    schema: {
        name: 'browser_tab_list',
        description: 'List browser tabs',
        inputSchema: (0, zod_to_json_schema_1.zodToJsonSchema)(zod_1.z.object({})),
    },
    handle: async (context) => {
        return {
            content: [{
                    type: 'text',
                    text: await context.listTabs(),
                }],
        };
    },
};
const selectTabSchema = zod_1.z.object({
    index: zod_1.z.number().describe('The index of the tab to select'),
});
const selectTab = captureSnapshot => ({
    capability: 'tabs',
    schema: {
        name: 'browser_tab_select',
        description: 'Select a tab by index',
        inputSchema: (0, zod_to_json_schema_1.zodToJsonSchema)(selectTabSchema),
    },
    handle: async (context, params) => {
        const validatedParams = selectTabSchema.parse(params);
        await context.selectTab(validatedParams.index);
        const currentTab = await context.ensureTab();
        return await currentTab.run(async () => {
            const code = [
                `// <internal code to select tab ${validatedParams.index}>`,
            ];
            return { code };
        }, { captureSnapshot });
    },
});
const newTabSchema = zod_1.z.object({
    url: zod_1.z.string().optional().describe('The URL to navigate to in the new tab. If not provided, the new tab will be blank.'),
});
const newTab = {
    capability: 'tabs',
    schema: {
        name: 'browser_tab_new',
        description: 'Open a new tab',
        inputSchema: (0, zod_to_json_schema_1.zodToJsonSchema)(newTabSchema),
    },
    handle: async (context, params) => {
        const validatedParams = newTabSchema.parse(params);
        await context.newTab();
        if (validatedParams.url)
            await context.currentTab().navigate(validatedParams.url);
        return await context.currentTab().run(async () => {
            const code = [
                `// <internal code to open a new tab>`,
            ];
            return { code };
        }, { captureSnapshot: true });
    },
};
const closeTabSchema = zod_1.z.object({
    index: zod_1.z.number().optional().describe('The index of the tab to close. Closes current tab if not provided.'),
});
const closeTab = captureSnapshot => ({
    capability: 'tabs',
    schema: {
        name: 'browser_tab_close',
        description: 'Close a tab',
        inputSchema: (0, zod_to_json_schema_1.zodToJsonSchema)(closeTabSchema),
    },
    handle: async (context, params) => {
        const validatedParams = closeTabSchema.parse(params);
        await context.closeTab(validatedParams.index);
        const currentTab = context.currentTab();
        if (currentTab) {
            return await currentTab.run(async () => {
                const code = [
                    `// <internal code to close tab ${validatedParams.index}>`,
                ];
                return { code };
            }, { captureSnapshot });
        }
        return {
            content: [{
                    type: 'text',
                    text: await context.listTabs(),
                }],
        };
    },
});
exports.default = (captureSnapshot) => [
    listTabs,
    newTab,
    selectTab(captureSnapshot),
    closeTab(captureSnapshot),
];
